File: test/args.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import {parseArgs} from "../src/args.ts";

test("parseArgs returns defaults", () => {
  const result = parseArgs([]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.all, false);
  assert.equal(result.options.maxDepth, 3);
  assert.equal(result.options.fullscreen, false);
  assert.equal(result.options.verbose, false);
  assert.equal(result.options.noFetch, false);
  assert.equal(result.options.help, false);
  assert.equal(result.options.version, false);
  assert.equal(result.options.sync, false);
  assert.equal(typeof result.options.dir, "string");
});

test("parseArgs supports flags", () => {
  const result = parseArgs(["--all", "--fullscreen", "--verbose", "--no-fetch", "--sync", "--help", "--version"]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.all, true);
  assert.equal(result.options.fullscreen, true);
  assert.equal(result.options.verbose, true);
  assert.equal(result.options.noFetch, true);
  assert.equal(result.options.sync, true);
  assert.equal(result.options.help, true);
  assert.equal(result.options.version, true);
});

test("parseArgs supports --max-depth value", () => {
  const result = parseArgs(["--max-depth", "5"]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.maxDepth, 5);
});

test("parseArgs supports --max-depth=value", () => {
  const result = parseArgs(["--max-depth=7"]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.maxDepth, 7);
});

test("parseArgs rejects invalid max depth", () => {
  assert.equal(parseArgs(["--max-depth", "0"]).ok, false);
  assert.equal(parseArgs(["--max-depth", "abc"]).ok, false);
  assert.equal(parseArgs(["--max-depth"]).ok, false);
});

test("parseArgs supports --dir value", () => {
  const result = parseArgs(["--dir", "/some/path"]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.dir, "/some/path");
});

test("parseArgs supports --dir=value", () => {
  const result = parseArgs(["--dir=/another/path"]);
  assert.equal(result.ok, true);
  if (!result.ok) return;
  assert.equal(result.options.dir, "/another/path");
});

test("parseArgs rejects missing --dir value", () => {
  assert.equal(parseArgs(["--dir"]).ok, false);
  assert.equal(parseArgs(["--dir", "--all"]).ok, false);
  assert.equal(parseArgs(["--dir="]).ok, false);
});

test("parseArgs rejects unknown args", () => {
  assert.equal(parseArgs(["--raw"]).ok, false);
});

File: internal/args/args_test.go
package args

import "testing"

func TestParseReturnsDefaults(t *testing.T) {
	result := Parse(nil, "/cwd")
	if !result.OK {
		t.Fatalf("Parse returned error: %v", result.Err)
	}
	if result.Options.MaxDepth != 3 {
		t.Fatalf("expected MaxDepth 3, got %d", result.Options.MaxDepth)
	}
	if result.Options.Dir != "/cwd" {
		t.Fatalf("expected Dir /cwd, got %s", result.Options.Dir)
	}
	if result.Options.Verbose || result.Options.NoFetch || result.Options.Sync || result.Options.Help || result.Options.Version {
		t.Fatalf("expected boolean flags to default false: %+v", result.Options)
	}
}

func TestParseSupportsFlags(t *testing.T) {
	result := Parse([]string{"--verbose", "--no-fetch", "--sync", "--help", "--version"}, "/cwd")
	if !result.OK {
		t.Fatalf("Parse returned error: %v", result.Err)
	}
	if !result.Options.Verbose || !result.Options.NoFetch || !result.Options.Sync || !result.Options.Help || !result.Options.Version {
		t.Fatalf("expected all flags true: %+v", result.Options)
	}
}

func TestParseSupportsMaxDepth(t *testing.T) {
	tests := []struct {
		name string
		argv []string
		want int
	}{
		{name: "separate value", argv: []string{"--max-depth", "5"}, want: 5},
		{name: "equals value", argv: []string{"--max-depth=7"}, want: 7},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			result := Parse(test.argv, "/cwd")
			if !result.OK {
				t.Fatalf("Parse returned error: %v", result.Err)
			}
			if result.Options.MaxDepth != test.want {
				t.Fatalf("expected MaxDepth %d, got %d", test.want, result.Options.MaxDepth)
			}
		})
	}
}

func TestParseRejectsInvalidMaxDepth(t *testing.T) {
	for _, argv := range [][]string{{"--max-depth", "0"}, {"--max-depth", "abc"}, {"--max-depth"}} {
		if Parse(argv, "/cwd").OK {
			t.Fatalf("expected Parse(%v) to fail", argv)
		}
	}
}

func TestParseSupportsDir(t *testing.T) {
	tests := []struct {
		argv []string
		want string
	}{
		{argv: []string{"--dir", "/some/path"}, want: "/some/path"},
		{argv: []string{"--dir=/another/path"}, want: "/another/path"},
	}

	for _, test := range tests {
		result := Parse(test.argv, "/cwd")
		if !result.OK {
			t.Fatalf("Parse returned error: %v", result.Err)
		}
		if result.Options.Dir != test.want {
			t.Fatalf("expected Dir %s, got %s", test.want, result.Options.Dir)
		}
	}
}

func TestParseRejectsMissingDir(t *testing.T) {
	for _, argv := range [][]string{{"--dir"}, {"--dir", "--verbose"}, {"--dir="}} {
		if Parse(argv, "/cwd").OK {
			t.Fatalf("expected Parse(%v) to fail", argv)
		}
	}
}

func TestParseRejectsUnknownArgs(t *testing.T) {
	for _, argv := range [][]string{{"--raw"}, {"--all"}} {
		if Parse(argv, "/cwd").OK {
			t.Fatalf("expected Parse(%v) to fail", argv)
		}
	}
}
